// vertex360/src/index.js
var superagent = require('superagent')
var Promise = require('bluebird')

// Vertex imports
var Vertex = require('./vertex/Vertex')
var Router = require('./vertex/Router')
var APIRouter = require('./vertex/APIRouter')
var Controller = require('./vertex/Controller')
var ConfigRouter = require('./vertex/ConfigRouter')
var utils = require('./vertex/utils')
var local = require('./vertex/local')
var db = require('./vertex/db')
// var graphQl = require('./graph-ql')

var Microservice = function(credentials){
	const BASE_URL = 'https://api.turbo360.co'
	const TURBO_URL = 'https://www.turbo360.co'

	var config = {
		site_id: credentials['site_id'],
		turbo_url: TURBO_URL, // for fetching turbo dashboard resources (site, profile, etc)
		base_url: BASE_URL
	}

	var vertexApp = function(opts){ // opts can be null
		const keys = Object.keys(process.env)
		keys.forEach(function(key, i) {
			if (key.indexOf('AWS_') != -1)
				delete process.env[key]
		})

		delete process.env['PATH']
		delete process.env['_HANDLER']
		delete process.env['LD_LIBRARY_PATH']
		delete process.env['LAMBDA_TASK_ROOT']
		delete process.env['LAMBDA_RUNTIME_DIR']

		if (process.env.TURBO_ENV == 'prod'){ // this is null in production
			var app = new Vertex(opts)
			return app
		}

		var app = local.app(opts)
		return app
	}

	var router = function(){
		if (process.env.TURBO_ENV == 'prod'){ // this is null in production
			var router = new Router()
			return router
		}

		// console.log('DEV ROUTER!')
		var router = local.router()
		return router
	}

	// var graphqlServer = function(){
	// 	return graphQl
	// }

	// middleware function for fetching global config:
  var fetchGlobal = function(apiKey, env, turboSDK){
    return function(req, res, next) {
  		req.appSlug = req.headers['Turbo-Vertex-App']
  		turboSDK.globalConfig(apiKey, env)
  		.then(function(globalConfig) {
  	    req.global = globalConfig
  			next()
  	  })
  		.catch(function(err) {
  			next()
  		})
  	}
  }

	var setConfig = function(env){
		return function(req, res, next) {
			req.config = {
				cdn: (env.TURBO_ENV=='dev') ? null : env.TURBO_CDN,
		    global: req.global || {} // global configuration for the site. see "global.json" file in "pages" directory
			}

			next()
		}
	}

	var client = {
		express: vertexApp,
		app: vertexApp,
		router: router,
    fetchGlobal: fetchGlobal,
		setConfig: setConfig,
		APIRouter: APIRouter,
		Controller: Controller,
		ConfigRouter: ConfigRouter,
		nedbConfig: db.nedbConfig,
		utils: utils
		// graphqlServer: graphqlServer
	}

	return client
}

module.exports = Microservice
