// vertex360/src/vertex/ConfigRouter.js
const express = require('express')
const path = require('path')
var Router = require('./Router')
const BASE_DIR = __dirname + '/../../../../'

class ConfigRouter {
  constructor(config){
    this.site_id = config.site_id
    this.api_key = config.api_key
    this.env = config.env || 'dev'
    this.tmp = config.tmp || '/tmp'

    this.turbo = require('turbo360')({site_id: config.site_id})
    this.createRouter = this.createRouter.bind(this)
    this.populateRouter = this.populateRouter.bind(this)
    this.local = this.local.bind(this)
    this.prod = this.prod.bind(this)
  }

  createRouter(env){
    return (env=='dev') ? new express.Router() : new Router()
  }

  populateRouter(router, handler){
    router.get('/:pagename', (req, res) => {
      handler(req.params.pagename, req)
    	.then(data => {
    		// res.json(data)
        res.data(data)
    	})
    	.catch(err => {
    		res.json({
    			confirmation: 'fail',
    			message: err.message
    		})
    	})
    })
  }

  local(handler){ // this is a standard express router
    const router = this.createRouter('dev')
    this.populateRouter(router, handler)
    return router
  }

  prod(handler){
    const router = this.createRouter('prod')
    this.populateRouter(router, handler)
    return router
  }

  router(handler){
    const router = this.createRouter(this.env)
    this.populateRouter(router, handler)
    return router
  }
}

module.exports = ConfigRouter
